### PR TITLE
Fix: dnslink domain resolving was broken; Add: no caching for those

### LINKED ISSUE #1234: When used to host static websites cache-timeout should not be 48 weeks
As a website owner hosting his website in IPFS I want to be able to make updates to that page.

# Scenarios

1. I use http://some.gate.way/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA/index.html to host the website
  - It's okay the page is cached for the next 48 weeks, when I'm doing updates I will change the hash in links
1. I use http://myweb.site/ with a `dnslink` TXT DNS entry
  - I will not change the URL itself but the hash in the DNS. Though the URL http://myweb.site/ is cached for the next 48 weeks and I'm not able to make changes because I'm not able to clear the visitors DNS caches

# Solution

Apply the solution used for `/ipns/...` hashs also to requests coming from a `dnslink` entry

# Steps to check
```bash
# curl -s -i -H 'Host: swagger.luzifer.io' http://localhost:8080/ | head -n20
HTTP/1.1 200 OK
Accept-Ranges: bytes
Cache-Control: public, max-age=29030400
Content-Length: 3921
Content-Type: text/html; charset=utf-8
Etag: index.html
Last-Modified: Thu, 01 Jan 1970 00:00:01 GMT
Suborigin: QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA
X-Ipfs-Path: /ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA/index.html/
Date: Sat, 16 May 2015 11:03:43 GMT

<!DOCTYPE html>
<html>
<head>
  <title>Swagger UI</title>
```

### PR DESCRIPTION
**Pull request by Luzifer**
fixes #1234 & #1267

@jbenet PTAL

### CODE REVIEW

**Comment by jbenet** [comment]
Thanks this is great. sorry for the resolve errors.

File: core/corehttp/ipns_hostname.go [hunk 1]
```diff
@@ -20,8 +21,11 @@ func IPNSHostnameOption() ServeOption {
 			defer cancel()
 
 			host := strings.SplitN(r.Host, ":", 2)[0]
-			if p, err := n.Namesys.Resolve(ctx, host); err == nil {
-				r.URL.Path = path.Join(p.String(), r.URL.Path)
+			if len(host) > 0 && isd.IsDomain(host) {
+				name := "/ipns/" + host
+				if _, err := n.Namesys.Resolve(ctx, name); err == nil {
+					r.URL.Path = path.Join("/ipns/", host) + r.URL.Path
```
Comment (jbenet) [inline core/corehttp/ipns_hostname.go]: maybe: `name + r.URL.Path` ?

**Comment by wking** [comment]
And we probably want to some tests with a mock DNS resolver (like
namesys/dns_test.go) to make sure this doesn't break again.  I think
that means we want a public API in namesys to manage the resolver (and
probably also the publisher) maps in mpns.  That's a bit awkward at
the moment, because the ‘resolver’ interface is internal.  Maybe
namesys needs public backend APIs (currently the mpns type, the
resolve helper, and the resolver interface) along side it's current
public frontent APIs (the Resolver, Publisher, and NameSys
interfaces).

**Comment by Luzifer** [comment]
Just using `path.Join(p.String, r.URL.Path)` leads to bad results… `path.Join` removes any trailing slash which then leads to redirect foo…

**Comment by Luzifer** [comment]
About tests: @jbenet and me had an idea for this: https://botbot.me/freenode/ipfs/2015-05-21/?msg=39768745&page=5

**Comment by wking** [comment]
On Thu, May 21, 2015 at 10:34:41AM -0700, Knut Ahlers wrote:
> Just using `path.Join(p.String, r.URL.Path)` leads to bad results…
> `path.Join` removes any trailing slash which then leads to redirect
> foo…

Ah, that makes sense.  All these little wrinkles to trip us up :p.

**Comment by wking** [comment]
On Thu, May 21, 2015 at 10:40:14AM -0700, Knut Ahlers wrote:
> About tests: @jbenet and me had an idea for this:
> https://botbot.me/freenode/ipfs/2015-05-21/?msg=39768745&page=5

I just see “we need to test this” there, not how to get a mock DNS for
sharness tests.  Personally, I'd be fine with some Go tests for this
functionality, since it's a lot easier to plug in a mock DNS resolver
in Go than it is to set one up for sharness tests.

**Comment by Luzifer** [comment]
We have several tests for the resolver stuff but none of those caught the feature to be fully destroyed. A full-stack-test would test the whole stack and check the "outside" functionality is still intact and not only parts of the codes are okay.

**Comment by whyrusleeping** [comment]
we should have a few domains on ipfs.io, like `test.ipfs.io` or `test1.ipfs.io` that have TXT records we can use for testing like this

**Comment by wking** [comment]
On Thu, May 21, 2015 at 11:14:21AM -0700, Jeromy Johnson wrote:
> we should have a few domains on ipfs.io, like `test.ipfs.io` or
> `test1.ipfs.io` that have TXT records we can use for testing like
> this

Tests that require a working network connection between the location
running the tests (Travis?) and some external entity (the DNS chain
leading to test.ipfs.io) make me jumpy.  For one thing, you could
never change the values in those TXT records without breaking
bisection for commits that assumed the old entries/values.  It's easy
enough to mock the DNS entries in Go, so while I like the security of
end-to-end sharness tests, I don't think they're worth the trouble in
this case, and I'd rather have in-Go tests for this functionality.

**Comment by jbenet** [comment]
> we should have a few domains on ipfs.io, like test.ipfs.io or test1.ipfs.io that have TXT records we can use for testing like this

Yep, absolutely. if people drop some records here, i'll add them.

> Tests that require a working network connection between the location
> running the tests (Travis?) and some external entity (the DNS chain
> leading to test.ipfs.io) make me jumpy.  For one thing, you could
> never change the values in those TXT records without breaking
> bisection for commits that assumed the old entries/values.  It's easy
> enough to mock the DNS entries in Go, so while I like the security of
> end-to-end sharness tests, I don't think they're worth the trouble in
> this case, and I'd rather have in-Go tests for this functionality.

Agreed with this too. I think we should have a test that is usually mocked, but the mocking can be turned off and run live when in the internet (on CI). (good to test for real).

**Comment by jbenet** [comment]
@Luzifer should i merge this, or wait for tests?

**Comment by Luzifer** [comment]
@jbenet for me it would really nice to have this live and working `dnslink` feature. That way I could continue putting documentation to IPFS… You're the project owner so you have to decide whether to wait or merge…

**Comment by jbenet** [comment]
@Luzifer i'd also want it in-- so that resolve actually works. 

hmmm, @Luzifer would you submit a follow-up PR? not sure how long it would take either way.

**Comment by Luzifer** [comment]
Actually I'd prefer someone with more knowledge about the whole project to write the tests…

**Comment by whyrusleeping** [comment]
@wking do you think you could handle writing a test or two for this?

**Comment by jbenet** [comment]
circlci's detecting races for us :)  cc @whyrusleeping 

```
==================
WARNING: DATA RACE
Read by goroutine 77:
  github.com/ipfs/go-ipfs/exchange/bitswap.(*Bitswap).ReceiveMessage()
      /home/ubuntu/.go_project/src/github.com/ipfs/go-ipfs/exchange/bitswap/bitswap.go:339 +0x395
  github.com/ipfs/go-ipfs/exchange/bitswap/testnet.(*networkClient).ReceiveMessage()
      <autogenerated>:48 +0xde
  github.com/ipfs/go-ipfs/exchange/bitswap/testnet.(*network).deliver()
      /home/ubuntu/.go_project/src/github.com/ipfs/go-ipfs/exchange/bitswap/testnet/virtual.go:75 +0x16c

Previous write by goroutine 97:
  github.com/ipfs/go-ipfs/exchange/bitswap.(*Bitswap).ReceiveMessage()
      /home/ubuntu/.go_project/src/github.com/ipfs/go-ipfs/exchange/bitswap/bitswap.go:339 +0x3b4
  github.com/ipfs/go-ipfs/exchange/bitswap/testnet.(*networkClient).ReceiveMessage()
      <autogenerated>:48 +0xde
  github.com/ipfs/go-ipfs/exchange/bitswap/testnet.(*network).deliver()
      /home/ubuntu/.go_project/src/github.com/ipfs/go-ipfs/exchange/bitswap/testnet/virtual.go:75 +0x16c

Goroutine 77 (running) created at:
  github.com/ipfs/go-ipfs/exchange/bitswap/testnet.(*network).SendMessage()
      /home/ubuntu/.go_project/src/github.com/ipfs/go-ipfs/exchange/bitswap/testnet/virtual.go:62 +0x265
  github.com/ipfs/go-ipfs/exchange/bitswap/testnet.(*networkClient).SendMessage()
      /home/ubuntu/.go_project/src/github.com/ipfs/go-ipfs/exchange/bitswap/testnet/virtual.go:89 +0xc9
  github.com/ipfs/go-ipfs/exchange/bitswap.(*Bitswap).send()
      /home/ubuntu/.go_project/src/github.com/ipfs/go-ipfs/exchange/bitswap/bitswap.go:446 +0x2db
  github.com/ipfs/go-ipfs/exchange/bitswap.(*Bitswap).taskWorker()
      /home/ubuntu/.go_project/src/github.com/ipfs/go-ipfs/exchange/bitswap/workers.go:74 +0x586
  github.com/ipfs/go-ipfs/exchange/bitswap.func·013()
      /home/ubuntu/.go_project/src/github.com/ipfs/go-ipfs/exchange/bitswap/workers.go:40 +0x6f
  github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/jbenet/goprocess.func·005()
      /home/ubuntu/.go_project/src/github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/jbenet/goprocess/impl-mutex.go:110 +0xa7

Goroutine 97 (running) created at:
  github.com/ipfs/go-ipfs/exchange/bitswap/testnet.(*network).SendMessage()
      /home/ubuntu/.go_project/src/github.com/ipfs/go-ipfs/exchange/bitswap/testnet/virtual.go:62 +0x265
  github.com/ipfs/go-ipfs/exchange/bitswap/testnet.(*networkClient).SendMessage()
      /home/ubuntu/.go_project/src/github.com/ipfs/go-ipfs/exchange/bitswap/testnet/virtual.go:89 +0xc9
  github.com/ipfs/go-ipfs/exchange/bitswap.(*Bitswap).send()
      /home/ubuntu/.go_project/src/github.com/ipfs/go-ipfs/exchange/bitswap/bitswap.go:446 +0x2db
  github.com/ipfs/go-ipfs/exchange/bitswap.(*Bitswap).taskWorker()
      /home/ubuntu/.go_project/src/github.com/ipfs/go-ipfs/exchange/bitswap/workers.go:74 +0x586
  github.com/ipfs/go-ipfs/exchange/bitswap.func·013()
      /home/ubuntu/.go_project/src/github.com/ipfs/go-ipfs/exchange/bitswap/workers.go:40 +0x6f
  github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/jbenet/goprocess.func·005()
      /home/ubuntu/.go_project/src/github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/jbenet/goprocess/impl-mutex.go:110 +0xa7
==================
```

**Comment by jbenet** [comment]
(let's do a test as a separate PR -- https://github.com/ipfs/go-ipfs/issues/1289)

**Pull request merged**